presto-main/src/main/java/io/prestosql/operator/scalar/LuhnCheckFunction.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.operator.scalar;
+
+import io.airlift.slice.Slice;
+
+import io.prestosql.spi.function.Description;
+import io.prestosql.spi.function.ScalarFunction;
+import io.prestosql.spi.function.SqlType;
+import io.prestosql.spi.type.StandardTypes;
+
+import static java.lang.Integer.parseInt;
+
+public final class LuhnCheckFunction
+{
+    private LuhnCheckFunction() {}
+
+    @Description("Returns true if the string is a valid credit card number and false if it is not")
+    @ScalarFunction("luhn_check")
+    @SqlType(StandardTypes.BOOLEAN)
+    public static boolean LuhnCheck(@SqlType(StandardTypes.VARCHAR) Slice slice)
+    {
+        int nDigits = slice.length();
+        int sum = 0;
+        boolean secondNumber = false;
+        String cardNo = slice.toStringUtf8();
+
+        for (int i = nDigits - 1; i >= 0; i--) {
+            int n = parseInt(cardNo.substring(i, i + 1));
```

Comment:
String#substring is expensive.

But you do not need String at all.
See below for exaple how to proces Slice varchar character-by-character:
https://github.com/prestosql/presto/blob/6d9d47e0909b3fe9367584ba8450827dbbb8e1d7/presto-main/src/main/java/io/prestosql/type/LikeFunctions.java#L134

Reply:
We can process the bytes of the slice. If they aren't in the range of ASCII digits, then the function throws. This handles any non-ASCII or non-BMP automatically.

Reply:
Changed this to "int n = getCodePointAt(slice, i) - 48;"  
And added a catch for InvalidUtf8Exception.